Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeRemoval.php

```diff
@@ -83,6 +88,155 @@ protected function removeRelationRecursivelyFromDatabaseIncludingNonReferencedNo
+            $event->contentStreamIdentifier,
+            $event->nodeAggregateIdentifier,
+            $event->originDimensionSpacePoint
+        );
+        if (!$nodeRecord instanceof NodeRecord) {
+            throw EventCouldNotBeAppliedToContentGraph::becauseTheSourceNodeIsMissing(get_class($event));
+        }
+
+        // create a hierarchy relation to the restoration's parent
+        foreach ($event->affectedCoveredDimensionSpacePoints as $coveredDimensionSpacePoint) {
+            $parentNodeRecord = $this->projectionContentGraph->findParentNodeRecordByOriginInDimensionSpacePoint(
+                $event->contentStreamIdentifier,
+                $event->originDimensionSpacePoint,
+                $coveredDimensionSpacePoint,
+                $event->nodeAggregateIdentifier
+            );
+            if (!$parentNodeRecord instanceof NodeRecord) {
+                throw EventCouldNotBeAppliedToContentGraph::becauseTheTargetParentNodeIsMissing(get_class($event));
+            }
+            $position = $this->projectionContentGraph->determineHierarchyRelationPosition(
+                $parentNodeRecord->relationAnchorPoint,
+                null,
+                null,
+                $event->contentStreamIdentifier,
+                $coveredDimensionSpacePoint
+            );
+            $this->getDatabaseConnection()->executeStatement(
+                '
+            INSERT INTO ' . HierarchyRelationRecord::TABLE_NAME . ' (
+                  parentnodeanchor,
+                  childnodeanchor,
+                  `name`,
+                  position,
+                  dimensionspacepoint,
+                  dimensionspacepointhash,
+                  contentstreamidentifier
+                )
+                VALUES(
+                    :parentRelationAnchor,
+                    :childRelationAnchorPoint,
+                    :name,
+                    :position,
+                    :dimensionSpacePoint,
+                    :dimensionSpacePointHash,
+                    :contentStreamIdentifier
+                )',
+                [
+                    'dimensionSpacePoint' => json_encode($coveredDimensionSpacePoint),
+                    'dimensionSpacePointHash' => $coveredDimensionSpacePoint->hash,
+                    'contentStreamIdentifier' => (string)$event->contentStreamIdentifier,
+                    'originDimensionSpacePointHash' => $event->originDimensionSpacePoint->hash,
+                    'nodeAggregateIdentifier' => (string)$event->nodeAggregateIdentifier,
+                    'childRelationAnchorPoint' => (string)$nodeRecord->relationAnchorPoint,
+                    'name' => $nodeRecord->nodeName?->value,
+                    'parentRelationAnchor' => (string)$parentNodeRecord->relationAnchorPoint,
+                    'position' => $position
+                ]
+            );
+        }
+
+        // cascade to all descendants
+
+        $dimensionSpacePointsToJoin = '"' . implode(
```

Comment:
can we add a comment what we try to build up here (how this query part looks like); and why we do that?